sky/skylet/job_lib.py

```diff
@@ -537,32 +542,44 @@ def update_job_status(job_ids: List[int],
     if len(job_ids) == 0:
         return []
 
-    # TODO: if too slow, directly query against redis.
     ray_job_ids = [make_ray_job_id(job_id) for job_id in job_ids]
-
     job_client = _create_ray_job_submission_client()
 
-    # In ray 2.4.0, job_client.list_jobs returns a list of JobDetails,
-    # which contains the job status (str) and submission_id (str).
-    ray_job_query_time = time.time()
-    job_detail_lists: List['ray_pydantic.JobDetails'] = job_client.list_jobs()
-
-    job_details = {}
-    ray_job_ids_set = set(ray_job_ids)
-    for job_detail in job_detail_lists:
-        if job_detail.submission_id in ray_job_ids_set:
-            job_details[job_detail.submission_id] = job_detail
-
     statuses = []
     for job_id, ray_job_id in zip(job_ids, ray_job_ids):
         # Per-job status lock is required because between the job status
         # query and the job status update, the job status in the databse
         # can be modified by the generated ray program.
         with filelock.FileLock(_get_lock_path(job_id)):
             status = None
-            if ray_job_id in job_details:
-                ray_status = job_details[ray_job_id].status
-                status = _RAY_TO_JOB_STATUS_MAP[ray_status]
+            job_record = _get_jobs_by_ids([job_id])[0]
+            original_status = job_record['status']
+            job_submitted_at = job_record['submitted_at']
+
+            ray_job_query_time = time.time()
+            if original_status == JobStatus.INIT:
+                if (job_submitted_at >= psutil.boot_time() and job_submitted_at
+                        >= ray_job_query_time - _PENDING_SUBMIT_GRACE_PERIOD):
+                    # The job id is reserved, but the job is not submitted yet.
+                    # We should keep it in INIT.
+                    status = JobStatus.INIT
+                else:
+                    # The job id is reserved, but stale.
+                    status = JobStatus.FAILED
+
+            try:
+                # Querying status within the lock is safer than querying
+                # outside, as it avoids the race condition when job table is
+                # updated after the ray job status query.
+                # Also, getting per-job status is faster than querying all jobs,
```

Comment:
IIRC previously we tried per-job query and found it to be slow. Could we do some quick avg case + lots-of-jobs timing tests?

Reply:
See the PR description for the timing of submitting lots-of-jobs.

We have removed this update from the critical job submission path and the update will only be called by the  so that it won't be an issue. 

For 124 jobs in ray job:
* get_jobs (over 5 times): 0.038
* individually checking
  * 1 job: 0.006
  * 10 jobs: 0.066
  * 50 jobs: 0.324
Script for testing:
```python
from sky.skylet import job_lib
import time
client = job_lib._create_ray_job_submission_client()
total = 0
for i in range(5):
  start = time.time()
  jobs = client.list_jobs()
  #for j in range(20, 21):
  #  try:
  #    client.get_job_status(f'{j}-ubuntu')
  #  except RuntimeError:
  #      pass
  end = time.time()
  print(len(jobs))
  total += end-start
  print(end-start)
print('avg', total / 5)
```